README.md
<img align="right" width="131" height="225" src="assets/mojito.png?raw=true">

# Mojito [![Build Status](https://circleci.com/gh/appcues/mojito.svg?style=svg)](https://circleci.com/gh/appcues/mojito) [![Docs](https://img.shields.io/badge/api-docs-green.svg?style=flat)](https://hexdocs.pm/mojito/Mojito.html) [![Hex.pm Version](http://img.shields.io/hexpm/v/mojito.svg?style=flat)](https://hex.pm/packages/mojito) [![License](https://img.shields.io/hexpm/l/mojito.svg)](https://github.com/appcues/mojito/blob/master/LICENSE.md)

<!-- MDOC !-->

Mojito is an easy-to-use, high-performance HTTP client built using the
low-level [Mint library](https://github.com/ericmj/mint).

Mojito is built for comfort _and_ for speed.  Behind a simple and
predictable interface, there is a sophisticated connection pool manager
that delivers maximum throughput with no intervention from the user.

Just want to make one request and bail?  No problem.  Mojito can make
one-off requests as well, using the same process-less architecture as
Mint.

## Quickstart

```elixir
{:ok, response} = Mojito.request(method: :get, url: "https://github.com")
```

## Why Mojito?

Mojito addresses the following design goals:

* _Little or no configuration needed._  Use Mojito to make requests to as
  many different destinations as you like, without thinking about
  starting or selecting connection pools.  Other clients like
  [Hackney](https://github.com/benoitc/hackney)
  (and [HTTPoison](https://github.com/edgurgel/httpoison)),
  [Ibrowse](https://github.com/cmullaparthi/ibrowse) (and
  [HTTPotion](https://github.com/myfreeweb/httpotion)), and
  Erlang's built-in [httpc](http://erlang.org/doc/man/httpc.html)
  offer this feature, except that...

* _Connection pools should be used only for a single destination._
  Using a pool for making requests against multiple destinations is less
  than ideal, as many of the connections need to be reset before use.
  Mojito assigns requests to the correct pools transparently to the user.
  Other clients, such as [Buoy](https://github.com/lpgauth/buoy), Hackney/
  HTTPoison, Ibrowse/HTTPotion, etc. force the user to handle this
  themselves, which is often inconvenient if the full set of HTTP
  destinations is not known at compile time.

* _Redundant pools to reduce GenServer-related bottlenecks._  Mojito can
  serve requests to the same destination from more than one connection
  pool, and those pools can be selected by round-robin at runtime in order
  to minimize resource contention in the Erlang VM.  This feature is
  unique to Mojito.

## Installation

Add `mojito` to your deps in `mix.exs`:

```elixir
{:mojito, "~> 0.7.10"}
```

## Configuration

The following `config.exs` config parameters are supported:

* `:timeout` (milliseconds, default 5000) -- Default request timeout.
* `:max_body_size` - Max body size in bytes. Defaults to nil in which
  case no max size will be enforced.
* `:transport_opts` (`t:Keyword.t`, default `[]`) -- Options to pass to
  the `:gen_tcp` or `:ssl` modules.  Commonly used to make HTTPS requests
  with self-signed TLS server certificates; see below for details.
* `:pool_opts` (`t:pool_opts`, default `[]`) -- Configuration options
  for connection pools.

The following `:pool_opts` options are supported:

* `:size` (integer) sets the number of steady-state connections per pool.
  Default is 5.
* `:max_overflow` (integer) sets the number of additional connections
  per pool, opened under conditions of heavy load.
  Default is 10.
* `:pools` (integer) sets the maximum number of pools to open for a
  single destination host and port (not the maximum number of total
  pools to open).  Default is 5.
* `:strategy` is either `:lifo` or `:fifo`, and selects which connection
  should be checked out of a single pool.  Default is `:lifo`.
* `:destinations` (keyword list of `t:pool_opts`) allows these parameters
  to be set for individual `:"host:port"` destinations.

For example:

```elixir
use Mix.Config

config :mojito,
  timeout: 2500,
  pool_opts: [
    size: 10,
    destinations: [
      "example.com:443": [
        size: 20,
        max_overflow: 20,
        pools: 10
      ]
    ]
  ]
```

Certain configs can be overridden with each request.  See `request/1`.

## Usage

Make requests with `Mojito.request/1` or `Mojito.request/5`:

```elixir
>>>> Mojito.request(:get, "https://jsonplaceholder.typicode.com/posts/1")
## or...
>>>> Mojito.request(%{method: :get, url: "https://jsonplaceholder.typicode.com/posts/1"})
## or...
>>>> Mojito.request(method: :get, url: "https://jsonplaceholder.typicode.com/posts/1")

{:ok,
 %Mojito.Response{
   body: "{\n  \"userId\": 1,\n  \"id\": 1,\n  \"title\": \"sunt aut facere repellat provident occaecati excepturi optio reprehenderit\",\n  \"body\": \"quia et suscipit\\nsuscipit recusandae consequuntur expedita et cum\\nreprehenderit molestiae ut ut quas totam\\nnostrum rerum est autem sunt rem eveniet architecto\"\n}",
   headers: [
     {"content-type", "application/json; charset=utf-8"},
     {"content-length", "292"},
     {"connection", "keep-alive"},
     ...
   ],
   status_code: 200
 }}
```

By default, Mojito will use a connection pool for requests, automatically
handling the creation and reuse of pools.  If this is not desired,
specify the `pool: false` option with a request to perform a one-off request.
See the documentation for `request/1` for more details.

## Self-signed SSL/TLS certificates

To accept self-signed certificates in HTTPS connections, you can give the
`transport_opts: [verify: :verify_none]` option to `Mojito.request`
or `Mojito.Pool.request`:

```elixir
>>>> Mojito.request(method: :get, url: "https://localhost:8443/")
{:error, {:tls_alert, 'bad certificate'}}

>>>> Mojito.request(method: :get, url: "https://localhost:8443/", opts: [transport_opts: [verify: :verify_none]])
{:ok, %Mojito.Response{ ... }}
```

## Telemetry

Mojito integrates with the standard
[Telemetry](https://github.com/beam-telemetry/telemetry) library.

See the [Mojito.Telemetry](https://github.com/appcues/mojito/blob/master/lib/mojito/telemetry.ex)
module for more information.

<!-- MDOC !-->

## Changelog

See the [CHANGELOG.md](https://github.com/appcues/mojito/blob/master/CHANGELOG.md).

## Contributing

Thanks for considering contributing to this project, and to the free
software ecosystem at large!

Interested in contributing a bug report?  Terrific!  Please open a [GitHub
issue](https://github.com/appcues/mojito/issues) and include as much detail
as you can.  If you have a solution, even better -- please open a pull
request with a clear description and tests.

Have a feature idea?  Excellent!  Please open a [GitHub
issue](https://github.com/appcues/mojito/issues) for discussion.

Want to implement an issue that's been discussed?  Fantastic!  Please
open a [GitHub pull request](https://github.com/appcues/mojito/pulls)
and write a clear description of the patch.
We'll merge your PR a lot sooner if it is well-documented and fully
tested.

Contributors and contributions are listed in the
[changelog](https://github.com/appcues/mojito/blob/master/CHANGELOG.md).
Heartfelt thanks to everyone who's helped make Mojito better.

## Copyright and License

Copyright 2018 Appcues, Inc.

This software is released under the [MIT License](./LICENSE.md).


CHANGELOG.md
# Changelog

## 0.7.10 (2021-10-27)

Fixed a bug around starting pools. Thanks,
[@reisub](https://github.com/reisub)!

Marked compatibility with Telemetry 1.0. Thanks,
[@jchristgit](https://github.com/jchristgit)!

## 0.7.9 (2021-07-20)

Improved docs. Thanks, [@kianmeng](https://github.com/kianmeng)!

## 0.7.8 (2021-07-16)

Fixed a few bugs around connection handling and chunk sizing. Thanks to
[@reisub](https://github.com/reisub), [@fahchen](https://github.com/fahchen),
[@bmteller](https://github.com/bmteller).

## 0.7.7 (2021-02-04)

Added Mojito.Telemetry. Thanks,
[@andyleclair](https://github.com/andyleclair)! And thanks to the
[Finch](https://github.com/keathley/finch) team, whose telemetry
implementation informed this one.

## 0.7.6 (2020-12-10)

Fixed a bug around HTTP/2 responses larger than 64kB. Thanks for the
reports, [@dch](https://github.com/dch) and
[@jayjun](https://github.com/jayjun)!

Reduced memory footprint of idle Mojito pools by forcing GC after
requests complete. Thanks for the reports,
[@axelson](https://github.com/axelson) and
[@hubertlepicki](https://github.com/hubertlepicki)!

## 0.7.5 (2020-11-06)

Fixed packaging bug in 0.7.4.

## 0.7.4 (2020-11-02)

Fixed handling of Mint error responses.
Thanks, [@alexandremcosta](https://github.com/alexandremcosta)!

Fixed a Dialyzer warning around keyword lists.
Thanks, [@Vaysman](https://github.com/Vaysman)!

## 0.7.3 (2020-06-22)

Moved core Mojito functions into separate `Mojito.Base` module for
easier interoperation with mocking libraries like Mox. Thanks,
[@bcardarella](https://github.com/bcardarella)!

## 0.7.2 (2020-06-19)

Fixed typespecs.

## 0.7.1 (2020-06-17)

Fixed bug where Mojito failed to correctly handle responses with
a `connection: close` header. Thanks,
[@bmteller](https://github.com/bmteller)!

## 0.7.0 (2020-06-17)

Added the `:max_body_size` option, to prevent a response body from
growing too large. Thanks, [@rozap](https://github.com/rozap)!

## 0.6.4 (2020-05-20)

Fixed bug where sending an empty string request body would hang certain
HTTP/2 requests. Thanks for the report,
[@Overbryd](https://github.com/Overbryd)!

## 0.6.3 (2020-03-17)

`gzip`ped or `deflate`d responses are automatically expanded by
Mojito. Thanks, [@mogorman](https://github.com/mogorman)!

The Freedom Formatter has been removed. `mix format` is now applied.

## 0.6.2 (2020-03-11)

Header values are now stringified on their way to Mint. Thanks,
[@egze](https://github.com/egze)!

Timeouts of `:infinity` are now supported. Thanks,
[@t8rsalad](https://github.com/t8rsalad)!

## 0.6.1 (2019-12-20)

Internal refactor to support different pool implementations. No features
were added or changed.

Code formatting improvements in docs. Thanks,
[@sotojuan](https://github.com/sotojuan)!

## 0.6.0 (2019-11-02)

Upgraded to Mint 1.0. Thanks, [@esvinson](https://github.com/esvinson)!

Fixed typo in CHANGELOG. Thanks, [@alappe](https://github.com/alappe)!

## 0.5.0 (2019-08-21)

Fixed bug where timed-out responses could arrive in connection with
the next request from that caller.  Thanks for the report and the
test case, [@seanedwards](https://github.com/seanedwards)!

Refactored to use `%Mojito.Request{}` structs more consistently across
internal Mojito functions.

## 0.4.0 (2019-08-13)

Upgraded to Mint 0.4.0.

Requests are automatically retried when we attempt to reuse a closed
connection.

Added `Mojito.Headers.auth_header/2` helper for formintg HTTP Basic
`Authorization` header.

Don't pass the URL fragment to Mint when making requests.
Thanks [@alappe](https://github.com/alappe)!

Improved examples and docs around making POST requests.
Thanks [@hubertlepicki](https://github.com/hubertlepicki)!

Removed noisy debug output.
Thanks for the report, [@bcardarella](https://github.com/bcardarella)!

## 0.3.0 (2019-05-08)

Major refactor.

All end-user requests pass through `Mojito.request/1`, which now
accepts keyword list input as well.  `Mojito.request/5` remains
as an alias, and convenience methods for `get/3`, `post/4`, `put/4`,
`patch/4`, `delete/3`, `head/3`, and `options/3` have been added
(thanks, [@danhuynhdev](https://github.com/danhuynhdev)!).

Connection pools are handled automatically, sorting requests to the
correct pools, starting pools when necessary, and maintaining
multiple redundant pools for GenServer efficiency.

## 0.2.2 (2019-04-26)

Fixed a bug where long requests could exceed the given timeout without
failing (#17).  Thanks for the report,
[@mischov](https://github.com/mischov)!

Improved documentation about receiving `:tcp` and `:ssl` messages.
Thanks for the report,
[@axelson](https://github.com/axelson)!

Removed an extra `Task` process creation in `Mojito.Pool.request/2`.

## 0.2.1 (2019-04-23)

Refactored `Mojito.request/5` so it doesn't spawn a process.  Now all
TCP messages are handled within the caller process.

Added `Mojito.request/1` and `Mojito.Pool.request/2`, which accept a
`%Mojito.Request{}` struct as input.

Removed dependency on Fuzzyurl in favor of built-in URI module.

## 0.2.0 (2019-04-19)

Messages sent by Mojito now contain a `:mojito_response` prefix, to allow
processes to select or ignore these messages with `receive`.
Thanks [@AnilRedshift](https://github.com/AnilRedshift)!

Upgraded to Mint 0.2.0.

## 0.1.1 (2019-03-28)

`request/5` emits better error messages when confronted with nil or blank
method or url.  Thanks [@AnilRedshift](https://github.com/AnilRedshift)!

## 0.1.0 (2019-02-25)

Initial release, based on [Mint](https://github.com/ericmj/mint) 0.1.0.



lib/xclient.ex
defmodule XClient do
  @moduledoc ~S"""
  XClient is a simplified HTTP client built using the
  low-level [XHTTP library](https://github.com/ericmj/xhttp).

  It provides an interface that will feel familiar to users of other
  Elixir HTTP client libraries.

  WARNING! This library currently depends on pre-release software (XHTTP).
  It is not yet recommended to use XClient in production.

  Currently only HTTP 1.x is supported; however, as support for HTTP 2.x
  in XHTTP is finalized, it will be added to XClient.

  ## Installation

  Add `xclient` to your deps in `mix.exs`:

      {:xclient, "~> 0.7.0-vendored-xhttp"}

  ## Single-request example

  `XClient.request/5` can be used directly for making individual
  requests:

      >>>> XClient.request(:get, "https://jsonplaceholder.typicode.com/posts/1")
      {:ok,
       %XClient.Response{
         body: "{\n  \"userId\": 1,\n  \"id\": 1,\n  \"title\": \"sunt aut facere repellat provident occaecati excepturi optio reprehenderit\",\n  \"body\": \"quia et suscipit\\nsuscipit recusandae consequuntur expedita et cum\\nreprehenderit molestiae ut ut quas totam\\nnostrum rerum est autem sunt rem eveniet architecto\"\n}",
         headers: [
           {"content-type", "application/json; charset=utf-8"},
           {"content-length", "292"},
           {"connection", "keep-alive"},
           ...
         ],
         status_code: 200
       }}

  ## Pool example

  `XClient.Pool.request/6` can be used when a pool of persistent HTTP
  connections is desired:

      >>>> children = [XClient.Pool.child_spec(MyPool)]
      >>>> {:ok, _pid} = Supervisor.start_link(children, strategy: :one_for_one)
      >>>> XClient.Pool.request(MyPool, :get, "http://example.com")
      {:ok, %XClient.Response{...}}

  Connection pooling in XClient is implemented using
  [Poolboy](https://github.com/devinus/poolboy).

  ## Self-signed SSL/TLS certificates

  To accept self-signed certificates in HTTPS connections, you can give the
  `transport_opts: [verify: :verify_none]` option to `XClient.request/5`
  or `XClient.Pool.request/6`:

      >>>> XClient.request(:get, "https://localhost:8443/")
      {:error, {:tls_alert, 'bad certificate'}}

      >>>> XClient.request(:get, "https://localhost:4443/", [], "", transport_opts: [verify: :verify_none])
      {:ok, %XClient.Response{...}}
  """

  alias XClient.{Error, Utils}

  @type headers :: [{String.t(), String.t()}]

  @type response :: %XClient.Response{
          status_code: pos_integer,
          headers: headers,
          body: String.t()
        }

  @type error :: %XClient.Error{
          reason: any,
          message: any
        }

  @type method :: :head | :get | :post | :put | :patch | :delete | :options

  @request_timeout Application.get_env(:xclient, :request_timeout, 5000)

  @doc ~S"""
  Performs an HTTP request and returns the response.

  Options:

  * `:timeout` - Response timeout in milliseconds.  Defaults to
    `Application.get_env(:xclient, :request_timeout, 5000)`.
  """
  @spec request(method, String.t(), headers, String.t(), Keyword.t()) ::
          {:ok, response} | {:error, error}
  def request(method, url, headers \\ [], payload \\ "", opts \\ []) do
    timeout = opts[:timeout] || @request_timeout

    with {:ok, pid} <- XClient.ConnServer.start_link(),
         :ok <- XClient.ConnServer.request(pid, self(), method, url, headers, payload, opts) do
      receive do
        reply ->
          GenServer.stop(pid)
          reply
      after
        timeout ->
          GenServer.stop(pid)
          {:error, %Error{reason: :timeout}}
      end
    end
    |> Utils.wrap_return_value()
  end
end
